c/cert/src/rules/EXP43-C/DoNotPassAliasedPointerToRestrictQualifiedParam.ql

```diff
@@ -0,0 +1,150 @@
+  result =
+    upperBound(e)
+        .minimum(min(Expr source | DataFlow::localExprFlow(source, e) | source.getValue().toInt()))
+}
+
+int getPointerArithmeticOperandStatedValue(CallToFunctionWithRestrictParametersArgExpr expr) {
+  result = getStatedValue(expr.(PointerArithmeticExpr).getOperand())
+  or
+  // edge-case: &(array[index]) expressions
+  result = getStatedValue(expr.(AddressOfExpr).getOperand().(PointerArithmeticExpr).getOperand())
+  or
+  // fall-back if `expr` is not a pointer arithmetic expression
+  not expr instanceof PointerArithmeticExpr and
+  not expr.(AddressOfExpr).getOperand() instanceof PointerArithmeticExpr and
+  result = 0
+}
+
+class PointerValueToRestrictArgConfig extends DataFlow::Configuration {
+  PointerValueToRestrictArgConfig() { this = "PointerValueToRestrictArgConfig" }
+
+  override predicate isSource(DataFlow::Node source) { pointerValue(source.asExpr()) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(CallToFunctionWithRestrictParameters call |
+      sink.asExpr() = call.getAPtrArg().getAChild*()
+    )
+  }
+}
+
+from
+  CallToFunctionWithRestrictParameters call, CallToFunctionWithRestrictParametersArgExpr arg1,
+  CallToFunctionWithRestrictParametersArgExpr arg2, int argOffset1, int argOffset2
+where
+  not isExcluded(call, Pointers3Package::doNotPassAliasedPointerToRestrictQualifiedParamQuery()) and
+  arg1 = call.getARestrictPtrArg() and
+  arg2 = call.getAPtrArg() and
+  arg1 != arg2 and
+  exists(PointerValueToRestrictArgConfig config, Expr source1, Expr source2 |
+    config.hasFlow(DataFlow::exprNode(source1), DataFlow::exprNode(arg1.getAChild*())) and
+    (
+      // one pointer value flows to both args
+      config.hasFlow(DataFlow::exprNode(source1), DataFlow::exprNode(arg2.getAChild*()))
+      or
+      // there are two separate values that flow from an AddressOfExpr of the same target
+      getAddressOfExprTargetBase(source1) = getAddressOfExprTargetBase(source2) and
+      config.hasFlow(DataFlow::exprNode(source2), DataFlow::exprNode(arg2.getAChild*()))
+    )
+  ) and
+  // get the offset of the pointer arithmetic operand (or '0' if there is none)
+  argOffset1 = getPointerArithmeticOperandStatedValue(arg1) and
+  argOffset2 = getPointerArithmeticOperandStatedValue(arg2) and
+  (
+    // case 1: the pointer args are the same.
+    // (definite aliasing)
+    argOffset1 = argOffset2
+    or
+    // case 2: the pointer args are different, a size arg exists,
+    // and the size arg is greater than the difference between the offsets.
+    // (potential aliasing)
+    exists(Expr sizeArg |
+      sizeArg = call.getAPossibleSizeArg() and
+      getStatedValue(sizeArg) > (argOffset1 - argOffset2).abs()
+    )
+    or
+    // case 3: the pointer args are different, and a size arg does not exist
+    // (potential aliasing)
+    not exists(call.getAPossibleSizeArg())
+  )
+select call,
+  "Call to '" + call.getTarget().getName() +
```

Comment:
I think the message should link to `arg2` as the aliased pointer, and link to `arg1` as the restrict-qualified parmeter. I think it should also provide a link to the source(s) (e.g. "derived from [source](source1)."  or "derived a pair of addressof expressions ([addressof1](source1), [addressof2](source2)) which take the address of [variable](getAddressOfExprTargetBase(source1)).")